server/sonar-server/src/main/java/org/sonar/server/es/textsearch/ComponentTextSearchFeatureRepertoire.java

```diff
@@ -53,35 +57,54 @@ public QueryBuilder getQuery(ComponentTextSearchQuery query) {
   },
   PREFIX(CHANGE_ORDER_OF_RESULTS) {
     @Override
-    public QueryBuilder getQuery(ComponentTextSearchQuery query) {
-      return prefixAndPartialQuery(query.getQueryText(), query.getFieldName(), SEARCH_PREFIX_ANALYZER)
+    public Stream<QueryBuilder> getQueries(ComponentTextSearchQuery query) {
+      List<String> tokens = split(query.getQueryText());
+      if (tokens.isEmpty()) {
+        return Stream.empty();
+      }
+      BoolQueryBuilder queryBuilder = prefixAndPartialQuery(tokens, query.getFieldName(), SEARCH_PREFIX_ANALYZER)
         .boost(3f);
+      return Stream.of(queryBuilder);
     }
   },
   PREFIX_IGNORE_CASE(GENERATE_RESULTS) {
     @Override
-    public QueryBuilder getQuery(ComponentTextSearchQuery query) {
+    public Stream<QueryBuilder> getQueries(ComponentTextSearchQuery query) {
       String lowerCaseQueryText = query.getQueryText().toLowerCase(Locale.getDefault());
-      return prefixAndPartialQuery(lowerCaseQueryText, query.getFieldName(), SEARCH_PREFIX_CASE_INSENSITIVE_ANALYZER)
+      List<String> tokens = split(lowerCaseQueryText);
+      if (tokens.isEmpty()) {
+        return Stream.empty();
+      }
+      BoolQueryBuilder queryBuilder = prefixAndPartialQuery(tokens, query.getFieldName(), SEARCH_PREFIX_CASE_INSENSITIVE_ANALYZER)
         .boost(2f);
+      return Stream.of(queryBuilder);
     }
   },
   PARTIAL(GENERATE_RESULTS) {
     @Override
-    public QueryBuilder getQuery(ComponentTextSearchQuery query) {
-      BoolQueryBuilder queryBuilder = boolQuery();
-      split(query.getQueryText())
+    public Stream<QueryBuilder> getQueries(ComponentTextSearchQuery query) {
+      List<String> tokens = split(query.getQueryText());
+      if (tokens.isEmpty()) {
+        return Stream.empty();
+      }
+      BoolQueryBuilder queryBuilder = boolQuery().boost(0.5f);
+      tokens.stream()
         .map(text -> tokenQuery(text, query.getFieldName(), SEARCH_GRAMS_ANALYZER))
         .forEach(queryBuilder::must);
-      return queryBuilder
-        .boost(0.5f);
+      return Stream.of(queryBuilder);
     }
   },
   KEY(GENERATE_RESULTS) {
     @Override
-    public QueryBuilder getQuery(ComponentTextSearchQuery query) {
-      return matchQuery(SORTABLE_ANALYZER.subField(query.getFieldKey()), query.getQueryText())
-        .boost(50f);
+    public Stream<QueryBuilder> getQueries(ComponentTextSearchQuery query) {
+      String queryText = query.getQueryText();
+      if (ComponentKeys.isValidModuleKey(queryText)) {
```

Comment:
Is it expected to not search for file keys anymore ? Here only projects and modules are supported.